### ISSUE TITLE
Add accessible descriptions and alerts or this sim

### ISSUE THREAD

**Issue by mbarlow12** [phetsims/faradays-law#109]
Jon Hung has created a fairly detailed mockup for the accessible text for the sim. Details can be found at:

https://wiki.fluidproject.org/display/fluid/PhET+Faraday%27s+Law+Simulation+-+Text+descriptions%2C+alerts%2C+and+keyboard+access

**Comment by mbarlow12** [comment]
More details on the text:

https://wiki.fluidproject.org/display/fluid/PhET+Faraday%27s+Law+Simulation+-+PDOM+Sketch

**Comment by mbarlow12** [comment]
New version on phet-dev:

https://phet-dev.colorado.edu/html/faradays-law/1.4.0-jump.2/phet/faradays-law_a11y_view.html

Please note that there's a bug in the a11y-view where pressing tab will move the sim in the iframe (it's been noted in an issue). Even so, navigation and interaction with the sim components still works.

After the above commit, the summary and dynamic content of the light bulb circuit are complete. Next will be finishing the static descriptions for the circuit and the magnet description content.

**Comment by mbarlow12** [comment]
I pushed a new dev release to phet-dev:

https://phet-dev.colorado.edu/html/faradays-law/1.4.0-jump.3/phet/faradays-law_en_phet.html?ea&accessibility

and a11y view

https://phet-dev.colorado.edu/html/faradays-law/1.4.0-jump.3/phet/faradays-law_a11y_view.html

Alerts are still on the way, and there are a couple lingering tweaks (e.g. pinning down the mapping for field strength and proximity read outs).

**Comment by mbarlow12** [comment]
New dev release with updated accessible order:

https://phet-dev.colorado.edu/html/faradays-law/1.4.0-jump.4/phet/faradays-law_a11y_view.html

**Comment by samreid** [comment]
@mbarlow12 said the preceding comment was addressed.

@jessegreenberg said we will need to synchronize our efforts since a11y is moving quickly and phet-io will need something stable for publication.

@samreid will set up set up a meeting for @jessegreenberg @mbarlow12 @zepumph and @samreid.

**Comment by mbarlow12** [comment]
@samreid the current implementation shows either arrows or arrows and the WASD keys depending on the presence of the `accessibility` query parameter (no parameter, no letter keys).

The interaction pattern is such that on mouse or keyboard drag, the interaction cue is removed until the sim is reset. This was also referenced by @zepumph in https://github.com/phetsims/faradays-law/issues/115, where he noted the way the arrows appear/disappear in

https://phet.colorado.edu/sims/html/faradays-law/latest/faradays-law_en.html.

**Comment by mbarlow12** [comment]
New dev version at:

https://phet-dev.colorado.edu/html/faradays-law/1.4.0-dev.45/phet/faradays-law_en_phet.html?ea&accessibility

**Comment by jbphet** [comment]
@emily-phet, @jessegreenberg and @terracoda - I came across this issue when working on the keyboard a11y feature.  The last comment on it was two years ago (to the day, which is kinda weird).  It wasn't assigned to anyone, and I'm not sure whether it should be closed or whether we should continue the description work here, so I thought I'd assign it to y'all to sort it out.

**Comment by emily-phet** [comment]
Thanks @jbphet 
Let's leave it this open, there are some useful links here for when we focus on description next time 'round. I'm assigning to myself only.

**Comment by terracoda** [comment]
@emily-phet, all the interactive description that has been implemented thus far is currently accessible using a screen reader.

Will this content go public automatically with publication? If so, I think we should quickly discuss pros and cons. If it is not going public en mass as-is, than there is no need to discuss at this time.

Items that are currently problematic and that might be trivial to add in for publication if content will be public by default.
1. No accessible name for the grab magnet button. VoiceOver currently says, Grab Object button.
    -  accessible name for the button should be: Grab Bar Magnet which would result in "Grab Bar Magnet button"
2. Issue with the utterance queue.
    - since this version of the sim is not being published with Interactive Description, I would suggest turning off the utterance queue or disabling Aria Live regions. Currently, there is an insane amount of pile ups of content being rad by VoiceOver. Again this is not an issue if the accessible PDOM content is not going live with publication.

**Comment by emily-phet** [comment]
@jbphet 

We discussed this question regarding whether or not description would be accessible in the published sim, and if there was an easy way to stop that from happening, a few weeks ago. 

I can't recall what we concluded...I suspect @jbphet might not have had a full answer on this and was perhaps going to discuss with @jessegreenberg.

From prior discussions re Ratio and Proportion, I believe it may be non-trivial to *not* have at least some of the description (perhaps the State descriptions) accessible upon publication. If that's the case, I think we should try to do @terracoda's suggestions above. 

- [ ] Adding an accessible name for the Bar Magnet
- [ ] Seeing if there is a way to "disable" all responses. 

Publishing the sim with what would be essentially a Scene summary and accessible names for the objects, etc. (description at a kind of "minimal" level) would be just fine.

What we definitely need to do here is minimize time spent on description at this time for this sim, so we're able to accomplish our remaining goals for this quarter within this quarter...so if one or both of these changes seem more time-consuming than somehow removing the existing description from the published sim, let's choose the least time-intensive option. 

@jessegreenberg can you help @jbphet sort this out?

Regardless, this sim will not be listed as having interactive description on the website, as its description is currently incomplete.

**Comment by jessegreenberg** [comment]
Since both alternative input and description (interactive descriptions) are implemented with the PDOM we currently don't have a way to publish one without the other. When this came up before we decided that it was OK to publish sims with full alternative input and incomplete description. Either way it will not be accessible to a screen reader user, and we didn't see the benefit of developing a way remove existing description in this case. All sims published with only alternative input have some incomplete descriptions that could be found if someone uses a screen reader.

Let me know if this needs to change and we can consider ways to support this.

**Comment by samreid** [comment]
Can the deferred label be removed?

**Comment by terracoda** [comment]
It would be supper nice to have an option to remove aria-live from a sim in a case like this, where all of the State Descriptions are essentially useful and correct. It's just the context responses which present a serious issue.

If we cannot easily turn off aria-live regions, another option is keep this sim as both a published sim and a prototype sim on the Accessible Sims page. The prototype version can include the Interactive Description icon, whereas the published version would not.

Once we start improving and finalizing the Interactive Description, new prototypes will become available through new prototypes.

**Comment by terracoda** [comment]
Yes, at the very least we should add an accessible name to the Bar Magnet interaction.

@jessegreenberg, do have an idea what it would take to create an option to “turn off” aria-live without breaking anything else in the PDOM?

**Comment by jessegreenberg** [comment]
Discussed with @emily-phet and @terracoda on 11/2/20 - 

We can add a way to disable alerts, which could be done with a query parameter and with a package.json entry, but we decided to keep the current categorization with enabling all screen reader features with the single `supportsInteractiveDescriptions`. Instead, we are going to comment out existing code that uses utterance-queue in the sim to prevent responses from piling up for the publication of this sim with alternative input.

**Comment by jbphet** [comment]
Undeferring so that the two items in https://github.com/phetsims/faradays-law/issues/109#issuecomment-717984131
 can be addressed.  This is necessary before publication of a version with keyboard nav and sound.

**Comment by jessegreenberg** [comment]
The sim-specific alerts have been removed for now so that they don't pile up if someone happens to use a screen reader in this version. I also added a label for the grabbable bar magnet.

@jbphet let me know if you have any questions, adding back the deferred label since we won't be working on descriptions any further for now.

**Comment by jbphet** [comment]
@emily-phet - I played around with this sim on a Mac with VoiceOver on after the changes made by @jessegreenberg, but unfortunately I don't have enough experience with description to know if the current behavior is what was envisioned for the next release.  Can you please check it out or have @terracoda take a look and let me know if we're good to go on this?  This issue is the last one to resolve before going to RC.

**Comment by jbphet** [comment]
I've assigned this to @terracoda at @emily-phet's request.

**Comment by terracoda** [comment]
@jbphet, so it looks like with a screen reader I read through the state descriptions (good). And when I tab through, I get object names on focus (good), and when I interact I get any native object responses, like native changes in states (e.g. checkbox and radiobutton selections) (good), AND then I receive NO custom responsive descriptions (i.e. alerts) that change the scene or give updates about position, or what's happening to light bulb, or that the sim is reset.

I think this all makes sense because it is the responsive descriptions that still need most of the work. The sonifications by themselves provide a lot of information, and can be heard without interference from incorrect or untimeley responsive descriptions.

The state descriptions, may need improvement, but they are not incorrect, and they are never untimely because they are accessed on demand.

I think this is great.

**Comment by terracoda** [comment]
and as far as we can go right now with this sim's features.

**Comment by terracoda** [comment]
I checked off items in https://github.com/phetsims/faradays-law/issues/109#issuecomment-717984131

**Comment by terracoda** [comment]
@jbphet, should I leave this issue open since the description is not totally done, but done for now?

**Comment by jbphet** [comment]
We're in a state where the partial description has been approved, so I'm going to start the process of publishing a new release.  This release will include sound, keyboard nav, pinch-and-zoom, and the partial description discussed above.  I'm going to mark this issue as deferred and we can reopen when we decide to implement more complete description.

Having said that, when we get back to working on description I'd recommend closing this issue and opening a new one given the size and scope to which this one has grown.

**Comment by terracoda** [comment]
@jbphet, I agree, I think it makes sense to close this issue and start with a fresh issue, once we pick up the descriptions again for this simulation. 

I'll go ahead and close :-)
